File: MIDILinker/MIDIDecoder.m
//
//  MIDIDecoder.m
//  MIDILink
//
//  Created by Julian Porter on 30/04/2016.
//  Copyright © 2016 JP Embedded Solutions. All rights reserved.
//

#import "MIDIDecoder.h"





@interface MIDIDecoder ()

@property (strong,atomic) NSLock *lock;

@end

@implementation MIDIDecoder

- (instancetype)init {
    self=[super init];
    if(self) {
        self.messages=[NSMutableArray array];
        self.timeStandard=[[TimeStandard alloc] init];
        self.lock=[[NSLock alloc] init];
    }
    return self;
}

+ (MIDIDecoder *)instance {
    return [[MIDIDecoder alloc] init];
}

- (void)loadPackets:(const MIDIPacketList *)packets {
    NSUInteger nPackets=packets->numPackets;
    if(nPackets>0) {
        const MIDIPacket *packet=&packets->packet[0];
        for(NSUInteger n=0;n<nPackets;n++) {
            MIDIMessage *message=[[MIDIMessage alloc] initFromPacket:packet withTimebase:self.timeStandard];
            [self.messages addObject:message];
            NSLog(@"%@",[message description]);
            
            packet=MIDIPacketNext(packet);
        }
        [[NSNotificationCenter defaultCenter] postNotificationName:@"MIDIDatatoDecode" object:nil];
    }
}

- (void)reset {
    [self.lock lock];
    [self.messages removeAllObjects];
    [self.lock unlock];
}

- (NSUInteger)count {
    [self.lock lock];
    NSUInteger n=[self.messages count];
    [self.lock unlock];
    return n;
}

- (NSString *) field:(NSString *)field ofRow:(NSInteger)row {
    NSString *s=nil;
    [self.lock lock];
    @try {
        if(row>=0 && row<[self.messages count]) {
            MIDIMessage *message=[self.messages objectAtIndex:row];
            s=[message valueForKey:field];
        }
    } @catch (NSException *exception) {
        s=nil;
    } @finally {}
    [self.lock unlock];
    return s;
}



@end

File: MIDITools/MIDIDecoder.swift
//
//  MIDIDecoder.swift
//  MIDITools
//
//  Created by Julian Porter on 06/04/2019.
//  Copyright © 2019 JP Embedded Solutions. All rights reserved.
//

import Foundation
import CoreMIDI

public typealias MIDIDecoderCallback = () -> ()

public class MIDIDecoder : Sequence  {
    public static let MIDIDataToDecode = Notification.Name("__MIDIDataToDecode")
    
    public struct Iterator : IteratorProtocol {
        public typealias Element = MIDIMessage
        
        private let messages : [MIDIMessage]
        public var it : Array<MIDIMessage>.Iterator
        
        public init(_ decoder: MIDIDecoder) {
            messages=decoder.messages
            it=messages.makeIterator()
        }
        
        public mutating func next() -> Element? {
            return it.next()
        }
        
    }

    
    fileprivate var messages : [MIDIMessage]
    private var timeStandard : MIDITimeStandard
    
    public init() throws {
        messages=[]
        timeStandard = try MIDITimeStandard()
    }
    
    
    public func reset() {
        return DispatchQueue.main.sync {
            self.messages.removeAll()
        }
    }
    
    public func load(_ packets : MIDIPacketList) {
        let nPackets = packets.numPackets
        if nPackets>0 {
            var pkt=packets.packet
            var packet = UnsafeMutablePointer<MIDIPacket>(&pkt)
            (0..<nPackets).forEach { _ in
                let message = MIDIMessage(packet.pointee, timebase: timeStandard)
                messages.append(message)
                packet=MIDIPacketNext(packet)
            }
            NotificationCenter.default.post(name: MIDIDecoder.MIDIDataToDecode, object: nil)
        }
    }
    
    public func get(field: String, ofRow row: Int) -> String? {
        return DispatchQueue.main.sync {
            if (0..<self.count).contains(row) {
                return self.messages[row][field]
            }
            else { return nil }
        }
    }
    
    public var count : Int { return DispatchQueue.main.sync { return messages.count } }
    public __consuming func makeIterator() -> Iterator {
        return DispatchQueue.main.sync { return Iterator(self) }
    }
    
    
}
